File: app/src/main/java/de/exercicse/jrossbach/podcaster/feature/widget/ui/Presenter.java
package de.exercicse.jrossbach.podcast;
import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.observers.DisposableCompletableObserver;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;


public abstract class Presenter<ViewType> {

    CompositeDisposable compositeDisposable;
    protected ViewType view;

    public void attachView(ViewType view){
        this.view = view;
        compositeDisposable = new CompositeDisposable();
    }

    protected void subscribe(Completable completable, DisposableCompletableObserver disposableCompletableObserver){
        compositeDisposable.add(completable.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(disposableCompletableObserver));
    }

    protected void subscribe(Single<?> single, DisposableSingleObserver disposableSingleObserver){
        compositeDisposable.add(single.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(disposableSingleObserver));
    }

    public void detachView(){
        cancelSubscription();
        view = null;
    }

    private void cancelSubscription() {
        if(compositeDisposable != null && !compositeDisposable.isDisposed()) {
            compositeDisposable.clear();
            compositeDisposable.dispose();
        }
    }
}

File: app/src/main/java/de/exercicse/jrossbach/podcast/Presenter.kt
package de.exercicse.jrossbach.podcast

import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.observers.DisposableCompletableObserver
import io.reactivex.schedulers.Schedulers

abstract class Presenter<ViewType: Any> {

    private var compositeDisposable: CompositeDisposable? = null
    internal var view: ViewType? = null

    fun attachView(view: ViewType) {
        this.view = view
        compositeDisposable = CompositeDisposable()
    }

    fun subscribe(completable: Completable, disposableCompletableObserver: DisposableCompletableObserver) {
        compositeDisposable?.add(
            completable
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(disposableCompletableObserver))
    }

    fun <ReturnType> subscribe(single: Single<ReturnType>,
        onSuccess: (ReturnType) -> Unit,
        onError: (Throwable) -> Unit) {
        compositeDisposable?.add(
            single
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(onSuccess, onError)
        )
    }

    fun detachView() {
        cancelSubscription()
        view = null
    }

    private fun cancelSubscription() {
        compositeDisposable?.let {
            if (!it.isDisposed) {
                it.clear()
                it.dispose()
            }
        }
    }
}
